action viewer overlay

sylviameows opened the pull request:
adds a tool tip overlay to code chest menus so you don't have to hover over the reference book to see descriptions. this overlay is toggle able through the config and is disabled by default.

## Examples
> ![example1](https://github.com/DFOnline/CodeClient/assets/66442608/d569e031-7a35-44f5-a0ca-9e4985d76e1d)
> example 1: a standard code block with minimal description.

> ![example2](https://github.com/DFOnline/CodeClient/assets/66442608/a674e7f0-f2bd-47b5-8cac-cd738609e201)
> example 2: a standard code block with a large description.

> ![example3](https://github.com/DFOnline/CodeClient/assets/66442608/efedcf8a-63c8-4ab9-b35b-503610c3aa63)
> example 3: example of normal tooltips displaying over the action's description.

> ![example4](https://github.com/DFOnline/CodeClient/assets/66442608/250cd3ca-b748-45d9-8156-1c8da4160f02)
> example 4: example of a call function block displaying it's description. (only works if reference book is in your inventory)

GeorgeRNG commented on src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:
```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    public static ItemStack getReferenceBook() {
```
This should be a utility, not just for this feature. We should also probably construct a hypercube.actiondump.Action from the reference book, which would allow usage of the ref book in the parameters ghosts.
You can just move it outside (to Utility perhaps) but you don't need to construct an action from it

sylviameows replied:
converting the reference book to an action was going to be my first approach but i hadn't wanted mess with parsing all of that data. however, i have submitted a commit that adds a new class `hypercube.ReferenceBook` that can be used for that purpose in the future with parameters ghosts and also moved `ActionViewer#getReferenceBook()` into `location.Dev` as it felt fitting to go there.

GeorgeRNG commented on src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:
```diff
@@ -0,0 +1,184 @@
+        if (CodeClient.location instanceof Dev dev && dev.isInDev(position) && world.getBlockEntity(position) instanceof ChestBlockEntity) {
+            var signEntity = world.getBlockEntity(position.down().west());
+            if (signEntity instanceof SignBlockEntity sign) {
+                if (sign.getFrontText().getMessage(0, true).contains(Text.literal("CALL FUNCTION"))) {
```
We should use the reference book to cover everything the actiondump doesn't know, so new actions or people who don't have the actiondump at all

GeorgeRNG commented on src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:
```diff
@@ -0,0 +1,184 @@
+                    if (item == null) return null;
+                    var displayBook = item.copy();
+                    var name = displayBook.getName().copy().withColor(0xFFFFFF);
+                    displayBook.setCustomName(name); // continuity with actions
```
Actions have no colour since you don't have an actiondump with colors:
- Running just `/getactiondump` will achieve this.
- Running `/getactiondump section` will get an actiondump with colours, which is intented behaviour.
tl;dr, they are supposed to have color

GeorgeRNG commented on src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:
```diff
@@ -0,0 +1,184 @@
+                tooltipWidth = width;
+            }
+        }
+        Vector2ic vector = positioner.getPosition(context.getScaledWindowWidth(), context.getScaledWindowHeight(), x, y, tooltipWidth, tooltipHeight);
```
if it's going of the edge, such as a small screen size, we should probably make hovering over the tooltip (or where it normally) is move it into the screen

sylviameows replied:
going to look into this, also do you have any suggestions on how to handle tool tips that are too tall to display on screen? originally i was thinking of doing something with scaling but i couldn't think of a way to do that.

sylviameows commented:
this feature is much more polished now, and has some extra benefit with viewing code blocks with large action sizes (like noise blocks). I'd like to look into adding some sort of easing on the movements, but that probably wouldn't fit into CodeClient's style.

GeorgeRNG commented on src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:
```diff
@@ -0,0 +1,223 @@
+    public static boolean scroll(Screen screen, double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
+        if (!(screen instanceof GenericContainerScreen || screen instanceof CustomChestMenu)) return false;
+        if (isValid() && tall && hover) {
+            scroll -= 8 * (int) verticalAmount;
```
scroll is inverted and can't scroll to the very top or bottom

GeorgeRNG replied:
ok scrolling to the top or bottom is a problem with the large custom chest on big enough GUI scales, since that menu goes off the screen. We could ignore this, or account for it.

sylviameows replied:
changed scroll inversion to be a config option (as for me it felt more natural) and fixed the scrolling issue on the large custom chest.

RedVortexDev commented:
hey 👋 
this looks really good.. thinking that if the tooltip was pinned to top instead of centered (closer to args) it would look better?

sylviameows commented:
i've added tooltip alignment and scrolling inversion as preferences in the config, and changed the defaults to have the feature enabled, aligned to the top, and use normal scrolling.
